### PR TITLE
Added sample code - ODSA for Java Devs - Spring Boot 3 + Data JPA P2

### PR DESCRIPTION
**Pull request by juarezjuniorgithub**
@Kuassim 
Signed-off-by: Juarez Barbosa Junior juarez.barbosa@gmail.com

### CODE REVIEW

File: java/odsa-spring/src/main/resources/application.properties [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# https://www.oracle.com/database/technologies/getting-started-using-jdbc.html
+spring.datasource.url=<URL>
+spring.datasource.username=<USERNAME>
+spring.datasource.password=<PASSWORD>
+
+# Properties for UCP - Universal Connection Pool (UCP). Spring Boot 2.4.0 or above is required
+spring.datasource.driver-class-name=oracle.jdbc.OracleDriver
+spring.datasource.type=oracle.ucp.jdbc.PoolDataSource
+spring.datasource.oracleucp.connection-pool-name=connectionPoolName1
+spring.datasource.oracleucp.initial-pool-size=15
```
Comment (Kuassim) [inline java/odsa-spring/src/main/resources/application.properties:10]: Do you need that many initial connections for this demo use case?

File: java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+import java.util.Objects;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+
+@Entity
+@Table(name = "EMPLOYEE", schema = "ADMIN")
```
Comment (Kuassim) [inline java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java:32]: The ADMIN schema should not be used for user/developer resources. As our code samples show the best practices, I'd suggest using a different schema. There use to be an HR/HR schema for demo purposes but I am not sure it's still around on Autonomous DB. If not you can ask the user to create one using SQL Developer Web interface (available on ADB console).

**Review by Kuassim — CHANGES_REQUESTED** [review]

File: java/odsa-spring/script/ODSASpringBoot.sql [hunk 3]
```diff
@@ -0,0 +1,33 @@
+CREATE USER ODSA_USER IDENTIFIED BY <YOUR_PASSWORD>;
+GRANT DWROLE TO ODSA_USER;
+GRANT CREATE SESSION TO ODSA_USER;
```
Comment (gvenzl) [inline java/odsa-spring/script/ODSASpringBoot.sql:3]: This is redundant, as it is already provided by the `DWROLE`, see the [documentation](https://docs.oracle.com/en/cloud/paas/autonomous-database/adbsa/manage-users-privileges.html#GUID-50450FAD-9769-4CF7-B0D1-EC14B465B873).

File: java/odsa-spring/src/main/resources/application.properties [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+spring.datasource.oracleucp.min-pool-size=10
+spring.datasource.oracleucp.max-pool-size=30
+spring.datasource.oracleucp.connection-factory-class-name=oracle.jdbc.pool.OracleDataSource
+
+# Enable JPA / Hibernate debugging
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+logging.level.org.hibernate.SQL=DEBUG
+logging.level.org.hibernate.type.descriptor.sql.BasicBinder=TRACE
```
Comment (gvenzl) [inline java/odsa-spring/src/main/resources/application.properties:19]: Do we really need debugging for a demo script? One would assume that a developer understands when/how to turn on debugging.

**Review by gvenzl — CHANGES_REQUESTED** [review]

File: java/odsa-spring/src/main/resources/application.properties (hunk 1 above)
Comment (juarezjuniorgithub) [reply java/odsa-spring/src/main/resources/application.properties:10]: Hi Kuassi, spot on! Not at all, I copied that verbatim from your OCW/J1 deck. Changed it to the values below as this is for educational purposes only indeed, so it's fixed now. Thanks!

spring.datasource.oracleucp.initial-pool-size=1
spring.datasource.oracleucp.min-pool-size=1
spring.datasource.oracleucp.max-pool-size=2

**Review by Kuassim — APPROVED** [review]

File: java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java (hunk 2 above)
Comment (juarezjuniorgithub) [reply java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java:32]: fixed

File: java/odsa-spring/script/ODSASpringBoot.sql [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+GRANT UNLIMITED TABLESPACE TO ODSA_USER;
```
[the remark this answers is not in the document]
Comment (juarezjuniorgithub) [reply java/odsa-spring/script/ODSASpringBoot.sql:4]: That's fixed, thanks for your support and orientation.

File: java/odsa-spring/script/ODSASpringBoot.sql (hunk 3 above)
Comment (juarezjuniorgithub) [reply java/odsa-spring/script/ODSASpringBoot.sql:3]: This is fixed as well, thanks again for your support and the orientation/documentation.

File: java/odsa-spring/pom.xml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+			</plugin>
+
+
+			<!-- Azure App Service -->
+			<plugin>
+				<groupId>com.microsoft.azure</groupId>
+				<artifactId>azure-webapp-maven-plugin</artifactId>
+				<version>2.5.0</version>
+				<configuration>
+					<subscriptionId>cc5cb530-ce48-488b-a1a7-5a1fec12ed5e</subscriptionId>
```
[the remark this answers is not in the document]
Comment (juarezjuniorgithub) [reply java/odsa-spring/pom.xml:81]: Fixed

**Comment by juarezjuniorgithub** [comment]
@Kuassim @jeandelavarene the changes were performed as required - fix indentation.
Please review, thanks.

**Review by Kuassim — APPROVED** [review]

File: java/odsa-spring/src/main/resources/application.properties (hunk 4 above)
Comment (juarezjuniorgithub) [reply java/odsa-spring/src/main/resources/application.properties]: adjustment accepted, debugging properties removed. @Kuassim please review, thanks

File: java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
+/*
+  Copyright (c) 2021, 2022, Oracle and/or its affiliates.
```
Comment (jeandelavarene) [inline java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java:2]: Change year to 2024

File: java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/EmployeeNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+package com.oracle.dev.jdbc.odsa.spring;
+
+class EmployeeNotFoundException extends RuntimeException {
+
+  private static final long serialVersionUID = 7768215094663842819L;
```
Comment (jeandelavarene) [inline java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/EmployeeNotFoundException.java:26]: Please check if this is needed.

File: java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java (hunk 7 above)
Comment (juarezjuniorgithub) [reply java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java:2]: Ok, I did it for all classes with the license note and updated them to 2024 as you requested

Comment (juarezjuniorgithub) [reply java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/Employee.java]: @Kuassim please review, approve and notify Jean per the process (2nd level approval). Thanks!

File: java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/EmployeeNotFoundException.java (hunk 8 above)
Comment (juarezjuniorgithub) [reply java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/EmployeeNotFoundException.java]: @Kuassim please review, approve and notify Jean per the process (2nd level approval). Thanks!

**Review by Kuassim — APPROVED** [review]

File: java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/EmployeeNotFoundException.java (hunk 8 above)
Comment (Kuassim) [reply java/odsa-spring/src/main/java/com/oracle/dev/jdbc/odsa/spring/EmployeeNotFoundException.java]: Approved; up to @jeandelavarene

**Review by Kuassim — APPROVED** [review]

**Comment by jeandelavarene** [comment]
@juarezjuniorgithub please remove the target directory from your PR. You can add it to your .gitignore.
See this: https://stackoverflow.com/questions/991801/git-ignores-and-maven-targets
Also please upgrade the JDBC driver to 23.6.

**Comment by juarezjuniorgithub** [comment]
Hi @Kuassim I have followed Jean's suggestion and updated the driver version to 23.6.0.24.10. 
However, let's try to abide by our latest agreements, as asking recurrently to update the driver version before the 
very first release of a given code sample configures a dead-end situation, where it's being kept in a situation of 
open modification. Such updates (versions/releases) will be done as part of our agreed Quarterly QA Freshness Pass activity. 
I have also addressed the issues with the /target directory as requested. 
@Kuassim please approve. Thanks

File: java/odsa-spring/README.md [hunk 9]
```diff
@@ -0,0 +1,2 @@
+# odsa-spring
+[]()
```
Comment (jeandelavarene) [inline java/odsa-spring/README.md:2]: The README should at least mention what this code sample is about.

**Review by jeandelavarene — APPROVED** [review]

**Review by jeandelavarene — APPROVED** [review]

**Review by gvenzl — APPROVED** [review]

**Pull request merged**